### DISCUSSION TITLE
Active Power Limit

### DISCUSSION THREAD

**Discussion by vbreemen** [WillCodeForCats/solaredge-modbus-multi#683] [Q&A] [answered]
I have question about the SolarEdge active power limit and hope someone can help. 

In the MySolarEdge app you can read the inverter active power limit as a percentage. 
The value in both the app and delivered by the web API (powerLimit %) is actually below 100% and correct when the inverter is throttling. 

When reading the modbus Sunspec registers, all known (to me) registers for reading power limit percentages always stay on 100%.
I tried these from the sunspec-implementation-technical-note and Power Control Open Protocol for SolarEdge Inverters documents: 

Name                                                            Address dec / hex
I_Active_Power_Limit                                 61441 / F001
I_Dynamic_Active_Power_Limit               62242 / F322
I_Dynamic_Reactive_Power_Limit           62244 / F324
I_PowerReduce                                           61760 / F140 

They all always return 100%
Does anyone know what is the correct register, or the calculation to get the correct percentage?

**Comment by WillCodeForCats** [comment]
There isn't a calculation, the registers are a direct read.

Example in code for `61441`:
https://github.com/WillCodeForCats/solaredge-modbus-multi/blob/fbb019792e4dbbd19dfd09456062949cce6c5b07/custom_components/solaredge_modbus_multi/hub.py#L1059
https://github.com/WillCodeForCats/solaredge-modbus-multi/blob/fbb019792e4dbbd19dfd09456062949cce6c5b07/custom_components/solaredge_modbus_multi/number.py#L536

Decoded as a 16 bit unsigned integer, and the value is returned directly to HA. So if that value never changes, something else is happening or the inverter itself has a firmware bug.

**Reply by vbreemen** [reply]
Thanks for taking a look. 61441 (F001) is an integer register while the power limit in the cloud API looks to be a float, has a value with two decimals also when throttling.

I asked SolarEdge too but they aren't communicative.

**Reply by WillCodeForCats** [reply]
Yeah on modbus `61441` is only whole number steps between 0 to 100 %.

**Reply by vbreemen** [reply]
Serious net congestion here. I'm using the active power configuration options which reduces the power limit in relation to the grid voltage using a graph defined in registers F190-F1A6. That way I can keep the grid voltage just below 253 Volt without tripping a shutdown. Works pretty well and the API power limit also shows the trottling.
The nice thing about it is that it's only limiting when the power is too high and not always.

It's not essential to be able to read the actual limit in HA,  just intriguing.

**Reply by WillCodeForCats** [reply]
The integration does read those values (PwrVsVgX and PwrVsVgY), but they aren't exposed to Home Assistant. You can see them when you download the diagnostics file or with debug logging turned on.

If those are something you want to use I can work on adding entities in HA for them.

**Reply by vbreemen** [reply]
Thanks Seth. Not necessary for me, I can set them myself.

**Reply by DamijanD** [reply]
As it is related I will ask here. I believe that with 3.0.0 I somehow lost active_power_limit (solaredge_mb_i1_active_power_limit). HA says that "This entity is no longer being provided by the solaredge_modbus_multi integration." But I can see that it is still present in [sensor.py](https://github.com/WillCodeForCats/solaredge-modbus-multi/blob/fbb019792e4dbbd19dfd09456062949cce6c5b07/custom_components/solaredge_modbus_multi/sensor.py#L1456)
Any idea what happened?

**Reply by WillCodeForCats** [reply]
There are no changes to entities in 3.0.0.

**Reply by vbreemen** [reply]
Same thing happened here, and also lost more of these advanced power control entities in HA, like CosPhi etc.

**Reply by WillCodeForCats** [reply]
You can turn on debug logging and see the detection process, but there's not been any code changes in the integration that would affect those entities.

**Reply by WillCodeForCats** [reply]
If you look at the history for `number.py` that contains the active power limit entity, the last change was 8 months ago in v2.4.10 (https://github.com/WillCodeForCats/solaredge-modbus-multi/commits/main/custom_components/solaredge_modbus_multi/number.py).

Also keep in mind that if you ever remove and re-add the hub and/or integration, this will use defaults from the installed release instead of carrying over settings through an upgrade. These changes listed in the release notes.

**Reply by vbreemen** [reply]
It just became unavailable about two weeks ago. First thought I had was that something changed in the inverter.
I haven't found the time yet to dive into it.

**Reply by DamijanD** [reply]
I know that I experimentally enabled power control and then disabled it back. Is it possible that this somehow disabled these properties? What would HA get from integration if (for some reason) integration couldn't read value from modbus? Probably entity would still exist just value would be undefined?

**Reply by DamijanD** [reply]
@vbreemen this option is important for these "advanced" entities: Auto-Detect Additional Entities, when I enabled this all missing entities showed up. 
Again thx @WillCodeForCats

**Reply by vbreemen** [reply]
Works. Thanks!

**Reply by WillCodeForCats** [reply]
Please remember to read the release notes, since some defaults have changed for some options and entity names. Existing settings are preserved on upgrade, but if the integration or hub is re-added/removed the new defaults will apply.

https://github.com/WillCodeForCats/solaredge-modbus-multi/releases/tag/v3.0.0

**Comment by vbreemen** [comment, marked as answer]
I eventually found the answer on my original question about the powerLimit % in the MySolarEdge app when throttling and in hindsight it's pretty obvious.
It's not in a separate register but a calculated value. When the inverter is throttling (register 40407 value 5) , it's the actual power (register 40083) divided by the maximum power limit of the inverter (register 62220).

**Reply by android118paolobur** [reply]
Good evening. I have a SolarEdge 6000 inverter without meter. From documentation https://knowledge-center.solaredge.com/sites/kc/files/sunspec-implementation-technical-note.pdf I can't find the value of the register for maximum power limit of the inverter (register 62220).
Putting the 62220. I don't get a response (Pymodbus: Paolo_solaredge_hub: Error: device: 1 address: 62220 -> Modbus Error: [Input/Output] No response received after 3 retries, continue with next request). Could you help me find the correct register value? Thanks

**Comment by vbreemen** [comment]
It's a float32 value so you need to read two adjacent registers. On my SE3680 inverter it contains the value 3680.0. The maximum power the inverter is rated to. I'm reading it with libmodbus C code as a swapped Float32 value.
It probably contains 6000.0 for your inverter.

**Reply by vbreemen** [reply]
It's actually a writeable register. The fixed maximum power rating is in read-only Float32 register 62212. Also returning 3680.0 on my SE3680 inverter.

https://www.photovoltaikforum.com/core/attachment/88445-power-control-open-protocol-for-solaredge-inverters-pdf/  pages 12 and 13. Hex F304 and F30C registers.
You may have to first enable dynamic power control (f300) before being able to write to f30c, but haven't tried that myself.

**Reply by heresiarch** [reply]
I could confirm the solution, I work with iobroker. 
I have an SE10K-RWB48 Hybrid Solaredge inverter.
Set register F300 to 1 once to enable the function. **Do not write to F300 again unless you explicitly need to change its value.**
After that you could use F322 to set the dynamic limit in percentage of the maximum power your Inverter could deliver.
The default Timeout until the system goes back to 100% seems to be 60 seconds. I kept that default, and write within the timeout/2 the actual power limit I need.
The Technote explains it very well, the good thing is that there is no need for accessing the installer settings or setup app.
[https://knowledge-center.solaredge.com/sites/kc/files/application_note_power_control_configuration.pdf](https://knowledge-center.solaredge.com/sites/kc/files/application_note_power_control_configuration.pdf)

**Comment by TomasTerhaag** [comment]
Great thread — the "it's calculated from actual ÷ rated, not a register you read
back" conclusion saved me too; the setpoint read-back sits stuck at 100 on my
HD-Wave (SE4000H, fw 0004.0024.0025) even at 0 W output.

Adding the *write* side for anyone arriving here trying to actually curtail: the
legacy `0xF001` percentage register burps on HD-Wave (holds 2-4s, snaps back to
full, repeats) and I couldn't stabilise it. The dynamic register **`0xF322`**
(Float32, low-word-first, % of nameplate via 0xF30C) holds perfectly when
re-asserted every ~10s, with Advanced Power Control ON and Command Timeout ≥ the
refresh interval. Since these are dynamic commands they don't persist to flash,
so the periodic re-assert doesn't wear anything (consistent with the flash-wear
note in #926).

Writeup + a live test script:
https://gist.github.com/TomasTerhaag/aa6aa059fbedee6f49a7c010ac5c9393
